pkg/reco/recommenders.go

```diff
@@ -0,0 +1,173 @@
+package reco
+
+import (
+	"context"
+	"errors"
+	v1alpha1 "github.com/flipkart-incubator/ottoscalr/api/v1alpha1"
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	golog "log"
+	"math"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+)
+
+type RecommendationWorkflow interface {
+	Execute(ctx context.Context, wm WorkloadMeta) (*v1alpha1.HPAConfiguration, *v1alpha1.HPAConfiguration, *Policy, error)
+}
+
+type Recommender interface {
+	Recommend(ctx context.Context, wm WorkloadMeta) (*v1alpha1.HPAConfiguration, error)
+}
+
+type RecommendationWorkflowImpl struct {
+	recommender     Recommender
+	policyIterators map[string]PolicyIterator
+	logger          logr.Logger
+}
+
+type WorkloadMeta struct {
+	metav1.TypeMeta
+	Name      string
+	Namespace string
+}
+
+type RecoWorkflowBuilder RecommendationWorkflowImpl
+
+func (b *RecoWorkflowBuilder) WithRecommender(r Recommender) *RecoWorkflowBuilder {
+	if b.recommender == nil {
+		b.recommender = r
+		return b
+	}
+	golog.Println("Only one recommender must be added. There's already one configured so ignoring this one.")
+	return b
+}
+
+func (b *RecoWorkflowBuilder) WithPolicyIterator(name string, p PolicyIterator) *RecoWorkflowBuilder {
+	if b.policyIterators == nil {
+		b.policyIterators = make(map[string]PolicyIterator)
+		b.policyIterators[name] = p
+	} else if _, ok := b.policyIterators[name]; !ok {
+		b.policyIterators[name] = p
+	}
+	return b
+}
+
+func (b *RecoWorkflowBuilder) WithLogger(log logr.Logger) *RecoWorkflowBuilder {
+	var zeroValLogger logr.Logger
+	if b.logger == zeroValLogger {
+		b.logger = log
+	}
+	return b
+}
+
+func (b *RecoWorkflowBuilder) Build() RecommendationWorkflow {
+	var zeroValLogger logr.Logger
+	if b.logger == zeroValLogger {
+		b.logger = zap.New()
+	}
+	return &RecommendationWorkflowImpl{
+		recommender:     b.recommender,
+		policyIterators: b.policyIterators,
+		logger:          b.logger,
+	}
+}
+
+func NewRecommendationWorkflowBuilder() *RecoWorkflowBuilder {
+	return &RecoWorkflowBuilder{}
+}
+
+type MockRecommender struct {
```

Comment:
Can we remove MockRecommender from this file and move it to test file if its still needed?

Reply:
This will be removed.